implementation/.github/workflows/update-dependencies-from-metadata.yml

```diff
@@ -91,19 +108,62 @@ jobs:
       matrix:
         includes: ${{ fromJSON(needs.retrieve.outputs.metadata-json) }}
-    # Run job step if
+    # Run job step if BOTH:
+    #   (1) needs.get-compile-and-test.outputs.should-test = TRUE -> if there is a dependency/test directory in the buildpack
+    #   (2) needs.get-compile-and-test.outputs.should-compile = FALSE -> if there is NOT a dependency/actions/compile directory in the buildpack
+    #   AND:
+    #   (3) there is at least one new version to test
+    if: ${{ needs.retrieve.outputs.length > 0 && needs.get-compile-and-test.outputs.should-test == 'true' && needs.get-compile-and-test.outputs.should-compile == 'false' }}
+    runs-on: ubuntu-latest
+    steps:
+      - name: Check out code
+        uses: actions/checkout@v2
+
+      - name: Make Temporary Artifact Directory
+        id: make-outputdir
+        run: |
+          echo "::set-output name=outputdir::$(mktemp -d)"
+
+      # Download the tarball for testing if:
+      #   (1) dependency testing code is present in the buildpack directory
+      #   (2) URI in metadata.json is available
+      - name: Download upstream tarball (if not compiled)
+        if: ${{ matrix.includes.uri != ''  && needs.get-compile-and-test.outputs.should-test == 'true' }}
+        run: |
+          curl ${{ matrix.includes.uri }} --silent --location --output ${{ steps.make-outputdir.outputs.outputdir }}/dependency.tgz
+
+      # Test the dependency tarball if:
+      #   (1) dependency testing code is present in the buildpack directory
+      - name: Test Upstream Dependency
+        working-directory: dependency
+        if: ${{ needs.get-compile-and-test.outputs.should-test == 'true' }}
+        run: |
+          make test \
+            version="${{ matrix.includes.version }}" \
+            tarballPath="${{ steps.make-outputdir.outputs.outputdir }}/*.tgz"
+
+  compile:
+    name: Compile and Test Dependency
+    needs:
+      - retrieve
+      - get-compile-and-test
+    strategy:
+      matrix:
+        includes: ${{ fromJSON(needs.retrieve.outputs.compilation-json) }}
+    # Run job step if:
     #   (1) needs.get-compile-and-test.outputs.should-compile -> if there is a dependency/actions/compile directory in the buildpack
-    #   (2) needs.get-compile-and-test.outputs.should-test -> if there is a dependency/test directory in the buildpack
-    #   (3) See line 113 -> if the dependency checksum and URI field are empty, signaling they need to be calculated
-    if: ${{ needs.get-compile-and-test.outputs.should-compile == 'true' || needs.get-compile-and-test.outputs.should-test == 'true' }}
+    #   (2) OR needs.get-compile-and-test.outputs.should-test -> if there is a dependency/test directory in the buildpack
+    #   AND:
+    #   (3) there is at least one version to compile/test
+    if: ${{ needs.retrieve.outputs.compilation-length > 0 && (needs.get-compile-and-test.outputs.should-compile == 'true' || needs.get-compile-and-test.outputs.should-test == 'true') }}
```

Comment:
does this `if` need an `always() && ${{ ... }}`?

Reply:
I think it's fine because we're not evaluating whether jobs are `skipped` - all dependent jobs must have passed, which is fine for this `if` statement.

I could see value in adding in the `always()` for consistency and future-proofing, but I can also see value in keeping it out as it is redundant.

Reply:
What would the value be of an `always()`? here? I'm not sure I can see the need for it, if you could explain a bit more about that

Reply:
Since it has `retrieve` and `get-compile-and-test` as `needs` I think that covers our bases, no?